[Site Isolation] Document::isTopDocument should return false for document of subframe

szewai opened the pull request:
#### d85ddaf99d035414fda2f2974bee09751b6ad50e
<pre>
[Site Isolation] Document::isTopDocument should return false for document of subframe
<a href="https://bugs.webkit.org/show_bug.cgi?id=273510">https://bugs.webkit.org/show_bug.cgi?id=273510</a>
<a href="https://rdar.apple.com/127308766">rdar://127308766</a>

Reviewed by Alex Christensen.

Currently `Document::isTopDocument` can return true when `Document` is in a subframe, as `Document::topDocument` returns
`this` when main frame is remote. When main frame is remote, the document cannot be top-level document, as the top-level
document should be in the same process as the main frame. To fix this, this patch makes `Document::isTopDocument` return
true only when main frame is local and main frame&apos;s document is the same as `this` (there could be multiple documents
pointing to the same frame, but only the one in use by frame can be the top). This change will make sure
`Document::isSameSiteForCookies` return the correct value and cookie test passes when site isolation flag is enabled.

To avoid breakage, the patch adds `Document::isTopDocumentLegacy`, which implements the old behavior of
`Document::isTopDocument`. `Document::isTopDocumentLegacy` is used when site isolation is not enabled, and at places
where new `Document::isTopDocument` implementation would cause breakage. We will migrate those call sites to use
`Document::isTopDocument` later.

Test: http/tests/cookies/same-site/popup-cross-site-from-cross-origin-iframe.html.

* Source/WebCore/dom/Document.cpp:
(WebCore::Document::isTopDocument const):
* Source/WebCore/dom/Document.h:
(WebCore::Document::setUserDidInteractWithPage):
(WebCore::Document::userDidInteractWithPage const):
(WebCore::Document::isTopDocumentLegacy const):
(WebCore::Document::isTopDocument const): Deleted.

Canonical link: <a href="https://commits.webkit.org/278222@main">https://commits.webkit.org/278222@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e858e833f16e8d6333a4437e25654afb97f8532c

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/49889 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/29176 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/2170 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/53132 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/566 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/52190 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/35237 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/134 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/40706 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/51987 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/26750 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/42929 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/21831 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/24184 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/147 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/8259 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/46241 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/194 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/54713 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/24983 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/130 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/48096 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/26240 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/43116 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/47137 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/10944 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/27102 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/25970 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

charliewolfe commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -10363,7 +10377,7 @@ bool Document::hasElementWithPendingUserAgentShadowTreeUpdate(Element& element)
 bool Document::isSameSiteForCookies(const URL& url) const
 {
-    auto domain = isTopDocument() ? RegistrableDomain(securityOrigin().data()) : RegistrableDomain(siteForCookies());
+    auto domain = isTopDocumentBasedOnFrame() ? RegistrableDomain(securityOrigin().data()) : RegistrableDomain(siteForCookies());
```
Why can't this just check isMainFrame() on m_frame?

szewai replied:
Good question. I think the issue is multiple `Document`s can point to the same `Frame` (`Document`'s frame is not changed after construction), but `Frame` only points to one `Document` (`Frame`'s document can change via `LocalFrame::setDocument`, and `Frame` does not uniquely own `Document`). 
The existing behavior of `isTopDocument()` is to return true only if the document is in use by the main frame, so I just kept the existing behavior for the new function.

achristensen07 replied:
Why can't we just change the implementation of isTopDocument to do this main frame check for all callers?  I think that would be better, but the change log indicates there is breakage.  I'm interested to know more about the breakage.

szewai replied:
It turns out the breakage is caused by web process crash on assertion failure; the other places seem to work (not causing observable failure for http/tests/cookies tests), so I just update the patch as suggested.

szewai committed:
[Site Isolation] Document::isTopDocument should return false for document of subframe
https://bugs.webkit.org/show_bug.cgi?id=273510
rdar://127308766

Reviewed by Alex Christensen.

Currently `Document::isTopDocument` can return true when `Document` is in a subframe, as `Document::topDocument` returns
`this` when main frame is remote. When main frame is remote, the document cannot be top-level document, as the top-level
document should be in the same process as the main frame. To fix this, this patch makes `Document::isTopDocument` return
true only when main frame is local and main frame's document is the same as `this` (there could be multiple documents
pointing to the same frame, but only the one in use by frame can be the top). This change will make sure
`Document::isSameSiteForCookies` return the correct value and cookie test passes when site isolation flag is enabled.

To avoid breakage, the patch adds `Document::isTopDocumentLegacy`, which implements the old behavior of
`Document::isTopDocument`. `Document::isTopDocumentLegacy` is used when site isolation is not enabled, and at places
where new `Document::isTopDocument` implementation would cause breakage. We will migrate those call sites to use
`Document::isTopDocument` later.

Test: http/tests/cookies/same-site/popup-cross-site-from-cross-origin-iframe.html.

* Source/WebCore/dom/Document.cpp:
(WebCore::Document::isTopDocument const):
* Source/WebCore/dom/Document.h:
(WebCore::Document::setUserDidInteractWithPage):
(WebCore::Document::userDidInteractWithPage const):
(WebCore::Document::isTopDocumentLegacy const):
(WebCore::Document::isTopDocument const): Deleted.

Canonical link: https://commits.webkit.org/278222@main
